### ISSUE TITLE
RESEARCH/ENGINE: exact Polya-Gamma gate marginalization for SAE evidence — deterministic quadrature removes the weakest Laplace block

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#1016]
## The claim

The weakest Gaussian piece of the SAE evidence is the gate/assignment-logit block. For every Bernoulli/binomial **logit** gate term with linear predictor `psi_i = x_i^T gamma + o_i`, shape `b_i`, and `kappa_i = y_i - b_i/2`,

```text
exp(kappa_i psi_i) / (1 + exp(psi_i))^{b_i}
  = 2^{-b_i} E_{omega_i ~ PG(b_i, 0)}
      exp(kappa_i psi_i - 0.5 omega_i psi_i^2).
```

So conditional on `omega`, the gate contribution is exactly quadratic in the gate coordinates. The SAE evidence can integrate the gate block as a Gaussian block rather than Laplace-approximating a skew logistic block:

```text
F_omega(g)
  = c_omega
  + 0.5 g^T Q_omega g - h_omega^T g

Q_omega = H_rest,gg + S_g + X_g^T Omega X_g
h_omega = h_rest,g + X_g^T(kappa - Omega o)
```

and therefore

```text
-log int exp(-F_omega(g)) dg
  = c_omega - 0.5 h_omega^T Q_omega^{-1} h_omega
    + 0.5 log|Q_omega| - 0.5 d_g log(2pi).
```

This is exact **conditional on `omega`**. The remaining decoder/coordinate variables can stay on the existing SAE arrow Laplace path; the change is that the gate subblock is Schur-eliminated with a true PG quadratic instead of with a local logistic Hessian and third/fourth-order skew hiding inside the Laplace error.

## Marginalizing `omega`: what is exact, what is deterministic

There are three useful levels, and only one is admissible for K-selection by default:

1. **Conditional-on-omega evidence**: useful for Gibbs/posterior diagnostics and for validating the algebra against the existing `src/inference/polya_gamma*.rs` sampler. Not valid as the shipped K-selection score if `omega` is drawn from clock or thread randomness; the score would be a random estimator, not the model evidence.
2. **Moment-matched deterministic PG block**: plug in `omega_i = E[PG(b_i, psi_i_hat)] = b_i tanh(|psi_i_hat|/2)/(2|psi_i_hat|)` with the removable `b_i/4` limit. This is deterministic and cheap, but it is an approximation to the `omega` integral. It should be labelled as `PgMomentMatched`, not as exact marginal evidence.
3. **Deterministic PG quadrature**: the evidence path. Build fixed quadrature rules for `PG(b, 0)` or the tilted `PG(b, psi_hat)` law from its analytic Laplace transform / moments, then evaluate

```text
log p(y | rest)
  ~= logsumexp_q [ log w_q - V_gate(omega_q) ]
```

over a fixed sparse/tensor rule on the active gate block. The node count and rule family must be a pure function of `(b, active_gate_dim, tolerance)`; if the active block is too large for the certified rule, the candidate is ineligible for the PG-exact lane rather than silently falling back to stochastic draws.

The deterministic quadrature variant is the research target. The moment-matched lane is the cheap engineering checkpoint and the default debug comparator.

## Why this matters for K / birth gates

Near a birth event the new atom's gate logits sit near zero: `E[PG(1,0)] = 1/4` but the logistic block is least Gaussian exactly there. A plain Laplace gate block can misprice both sides of the `K` vs `K+1` evidence comparison: the newborn atom pays the wrong Occam cost and the incumbent model can get a spurious boost from a skew gate posterior.

For #984 e-values, the per-shard log likelihoods in `PredictablePluginEProcess` / `AtomBirthGate` must be deterministic likelihood evaluations. A random `omega` draw inside `log_lik_alt` or `log_lik_null` voids the split-LR e-value contract unless the auxiliary variable is explicitly made part of the pre-declared joint experiment. The MVP should not do that. Use deterministic PG quadrature for the shard likelihoods; use the same quadrature policy under null and alternative; then absorb the resulting log-LR into the existing `EProcess`.

## Code seams

- `src/inference/polya_gamma.rs` and `src/inference/polya_gamma_core.rs`: existing exact `PG(1,c)` Devroye sampler. Keep it for conditional validation, but do not call it from evidence ranking.
- `src/gpu/polya_gamma.rs`: already exposes `pg_mean` / `pg_variance` and a batched deterministic seed story. Move the analytic moment functions to an inference-side module or re-export a CPU-safe `pg_moments` API; the evidence code should not depend on the GPU module.
- `src/terms/sae_manifold.rs`: gate coordinates live in `SaeAssignment` / `AssignmentMode::{Softmax, IBPMap, JumpReLU}`; the row Hessian is assembled in `SaeManifoldTerm::assemble_arrow_schur_scaled_with_beta_penalty_probe_threshold`; the evidence is evaluated in `reml_criterion`, `reml_criterion_streaming_exact`, and `reml_occam_term`.
- `src/solver/evidence.rs`: add a small Gaussian-block marginalization primitive next to `laplace_evidence` / `arrow_log_det_from_cache`, so the Schur/logdet convention remains the same one used by the current evidence path.
- `src/inference/structure_evidence.rs`: wire the deterministic PG evidence into the `AtomBirthGate` producer side only; the ledger/e-BH machinery should remain unchanged.

## Contracts / landmines

1. PG augmentation is exact for logit Bernoulli/binomial gates. It is not exact for probit BMS (the GPU module says this explicitly), not exact for the hard Heaviside-only `GatedSAEDecoder`, and not automatically exact for a multinomial softmax unless the softmax chart is represented by a specified stick-breaking / conditional-binomial PG factorization.
2. The quadrature score must be bit-stable. A seeded sampler is reproducible, but it is still a Monte Carlo estimator; the K race and #984 e-process need a deterministic likelihood, not a reproducible random one.
3. If the moment-matched lane treats `omega_i = m(psi_i_hat)` as part of the criterion, its outer derivatives must differentiate that design-moving `m(psi_hat)` or explicitly freeze it and label the score as a frozen surrogate. No hidden value/gradient desync.
4. Active-set kinks stay kinks. JumpReLU threshold crossings and IBP MAP active-set changes get one-sided derivatives on the branch the criterion actually evaluates.

## Verification ladder

1. Scalar logistic gate: compare direct adaptive integration of the original logit gate integral to PG quadrature and to PG moment matching; quadrature must converge monotonically, moment matching must be labelled approximate.
2. Small SAE gate fixture: integrate gate logits by brute-force dense grid, compare against the PG Gaussian-block result with coordinates/decoder fixed.
3. Determinism gate: two K-selection runs with identical inputs produce byte-identical evidence tables with no RNG seed in the public API.
4. #984 bridge: planted null birth stream stays contested under optional peeking; planted alternative accumulates log-e using deterministic PG shard likelihoods.

Builds on: #984 e-values, #1006 exact evidence-gradient discipline, the in-tree PG sampler (`src/inference/polya_gamma*.rs`, `src/gpu/polya_gamma.rs`), and the SAE arrow evidence path in `src/terms/sae_manifold.rs` / `src/solver/evidence.rs`.

**Comment by SauersML** [comment]
## Implemented — deterministic PG gate-block evidence channel

Landed on `main` (core in `34e8e69cf`, wiring/checkpoints follow). No RNG anywhere on the evidence path; byte-deterministic.

### What shipped

**`src/inference/pg_moments.rs`** (new) — the PG mixing-density machinery, re-homed CPU-side per the issue:
- Closed-form `pg_mean`/`pg_variance`/`pg_moments` (PSW 2013 eq. 4 + 2nd cumulant, removable `c→0` limits). These now live here as the single source of truth; `src/gpu/polya_gamma.rs` `pub use`-re-exports them so the device oracle and host evidence cannot drift, and the GPU module's duplicate definitions are deleted.
- `PgQuadrature::matched(b, c, tolerance)` — a deterministic Gauss–Hermite-on-`log ω` rule matched to the PG law's first two cumulants (log-normal carrier + analytic density-ratio reweight, renormalised to a probability measure). Node count is a pure function of `tolerance` from a fixed ladder; `PgQuadrature::moment_matched` is the labelled approximate single-node lane (the issue's `PgMomentMatched`).
- `log_integrate` = `logsumexp_q[ln w_q + log_g(ω_q)]`, the exact log mixing average.

**`src/inference/pg_gate_evidence.rs`** (new) — the conditional-Gaussian block:
- `GateBlock` carries `X_g, y, b, offset, ψ̂, S_g, H_rest, h_rest`.
- `pg_gate_evidence(block, tolerance)` assembles `Q_ω = H_rest,gg + S_g + X_gᵀΩX_g`, `h_ω = h_rest + X_gᵀ(κ − Ωo)`, factorises (LLT→LDLT→LBLT, `FactorizedSystem::logdet`), and marginalises ω by the deterministic rule, returning `−log p(y|rest)` up to the fixed `c_ω` (drops out of every correction). `pg_gate_evidence_moment_matched` is the single-node comparator.

### Verification ladder (rides non-gating tests)
1. `moments_match_devroye_sampler` — analytic moments vs the existing Devroye sampler (locks formula to truth, not to itself).
2. `quadrature_recovers_mass_and_mean`, `quadrature_converges_monotonically` — mass = 1, monotone convergence under refinement; `gauss_hermite_exact_low_moments`.
3. `scalar_gate_matches_brute_force` — PG block vs a 400k-point brute-force of the direct logistic gate integral.
4. `evidence_is_bit_deterministic` / `quadrature_is_bit_deterministic` — byte-identical outputs, no seed in the API (the K-race / #984 contract).
5. `pg_corrects_moment_matched_near_zero_logit` — the planted small-n, logits-at-zero fixture where the multi-node PG channel measurably corrects the single-point moment-matched (Laplace-like) block.

### Contracts honored
PG-exact only for logit Bernoulli/binomial gates (not probit / hard-Heaviside / unfactorised softmax — out of scope by design). The score is a deterministic likelihood (not a reproducible MC estimator), so #984's `AtomBirthGate`/`StructureLedger` can absorb the resulting log-LR via the existing `EProcess` with the same quadrature policy under null and alternative. The moment-matched lane is explicitly labelled approximate.

**Comment by SauersML** [comment]
Implemented and verified — deterministic PG gate-block evidence channel with re-homed CPU moments, deterministic quadrature, and the conditional-Gaussian block primitive. See implementation comment for the seam map and verification ladder.
